nexus/src/app/background/common.rs

```diff
@@ -0,0 +1,854 @@
+            'a: 'c,
+            'b: 'c,
+        {
+            async {
+                let count = self.counter;
+                self.counter += 1;
+                self.tx.send_replace(count);
+                serde_json::Value::Number(serde_json::Number::from(count))
+            }
+            .boxed()
+        }
+    }
+
+    async fn wait_until_count(mut rx: watch::Receiver<usize>, count: usize) {
+        loop {
+            let v = rx.borrow_and_update();
+            assert!(*v <= count, "count went past what we expected");
+            if *v == count {
+                return;
+            }
+            drop(v);
+
+            tokio::time::timeout(Duration::from_secs(5), rx.changed())
+                .await
+                .unwrap()
+                .unwrap();
+        }
+    }
+
+    // Verifies that activation through each of the three mechanisms (explicit
+    // signal, timeout, or dependency) causes exactly the right tasks to be
+    // activated
+    #[nexus_test(server = crate::Server)]
+    async fn test_driver_basic(cptestctx: &ControlPlaneTestContext) {
+        let nexus = &cptestctx.server.apictx().nexus;
+        let datastore = nexus.datastore();
+        let opctx = OpContext::for_tests(
+            cptestctx.logctx.log.clone(),
+            datastore.clone(),
+        );
+
+        // Create up front:
+        //
+        // - three ReportingTasks (our background tasks)
+        // - two "watch" channels used as dependencies for these tasks
+
+        let (t1, rx1) = ReportingTask::new();
+        let (t2, rx2) = ReportingTask::new();
+        let (t3, rx3) = ReportingTask::new();
+        let (dep_tx1, dep_rx1) = watch::channel(0);
+        let (dep_tx2, dep_rx2) = watch::channel(0);
+        let mut driver = Driver::new();
+
+        assert_eq!(*rx1.borrow(), 0);
+        let h1 = driver.register(
+            "t1".to_string(),
+            Duration::from_millis(100),
+            Box::new(t1),
+            opctx.child(std::collections::BTreeMap::new()),
+            vec![Box::new(dep_rx1.clone()), Box::new(dep_rx2.clone())],
+        );
+
+        let h2 = driver.register(
+            "t2".to_string(),
+            Duration::from_secs(300), // should never fire in this test
+            Box::new(t2),
+            opctx.child(std::collections::BTreeMap::new()),
+            vec![Box::new(dep_rx1.clone())],
+        );
+
+        let h3 = driver.register(
+            "t3".to_string(),
+            Duration::from_secs(300), // should never fire in this test
+            Box::new(t3),
+            opctx,
+            vec![Box::new(dep_rx1), Box::new(dep_rx2)],
+        );
+
+        // Wait for four activations of our task.  (This is three periods.) That
+        // should take between 300ms and 400ms.  Allow extra time for a busy
+        // system.
+        let start = Instant::now();
+        let wall_start = Utc::now();
+        wait_until_count(rx1.clone(), 4).await;
+        assert_eq!(*rx1.borrow(), 4);
```

Comment:
I think that's right.  I loosened this check in ba037e099d7dfc022d9661f4c0cc5ae14dcc4baa.  I don't think the other checks are affected because the one at L633 already accounts for this and the other checks use a task with a long period that we expect won't fire during the test.